contracts/UnionZap.sol

```diff
@@ -319,64 +363,213 @@ contract UnionZap is Ownable, UnionBase {
+                prices[i] = ICurveV2Pool(tokenInfo[_outputToken].pool)
+                    .price_oracle();
+                // compute ETH value of current token balance
+                amounts[i] =
+                    (IERC20(_outputToken).balanceOf(address(this)) *
+                        prices[i]) /
+                    1e18;
+                // add the ETH value of token to current ETH value in contract
+                _totalEthBalance += amounts[i];
+            }
+        }
 
-        // transfer funds
-        IERC20(CVXCRV_TOKEN).safeTransfer(unionDistributor, _netDeposit);
-        if (stake) {
-            IMerkleDistributorV2(unionDistributor).stake();
+        // second loop to balance the amounts with buys and sells before distribution
+        for (uint256 i; i < weights.length; ++i) {
+            // if weight == 0, the token would have been swapped already so no balance
+            if (weights[i] > 0) {
+                _outputToken = outputTokens[i];
+                // amount adjustments
+                uint256 _desired = (_totalEthBalance * weights[i]) / DECIMALS;
+                if (amounts[i] > _desired) {
+                    _sell(
+                        _outputToken,
+                        (((amounts[i] - _desired) * 1e18) / prices[i])
+                    );
+                } else {
+                    _buy(_outputToken, _desired - amounts[i]);
+                }
+                // we need an edge case here since it's too late
+                // to update the cvxCRV distributor's stake function
+                if (_outputToken == CRV_TOKEN) {
+                    // convert all CRV to cvxCRV
+                    _toCvxCrv(minAmounts[i], lock);
+                } else {
+                    // slippage check
+                    assert(
+                        IERC20(_outputToken).balanceOf(address(this)) >
+                            minAmounts[i]
+                    );
+                }
+            }
         }
+    }
 
-        if (_gasCostInCrv > 0) {
-            uint256 _gasEth = _swapCrvToEth(
-                _swapCvxCrvToCrv(_gasCostInCrv, address(this))
-            );
-            (bool success, ) = (msg.sender).call{value: _gasEth}("");
-            require(success, "ETH transfer failed");
+    /// @notice Deposits rewards to their respective merkle distributors
+    /// @param weights - weights of output assets (cvxCRV, FXS, CVX...)
+    function distribute(uint16[] calldata weights)
+        public
+        onlyOwner
+        validWeights(weights)
+    {
+        for (uint256 i; i < weights.length; ++i) {
+            if (weights[i] > 0) {
+                address _outputToken = outputTokens[i];
+                address _distributor = tokenInfo[_outputToken].distributor;
+                IMerkleDistributorV2(_distributor).freeze();
+                // edge case for CRV as we gotta keep using existing distributor
+                if (_outputToken == CRV_TOKEN) {
```

Comment:
Could there be potential issues with `_distributor` being assigned to the distributor for `CRV_TOKEN`?

Reply:
I don't think there would unless we ever have another pounder option for CRV besides the cvxCRV one, but if it gets to this we'd have to rewrite the entire contract to handle multiple distributors for single assets anyways.